tensorflow/core/profiler/internal/gpu/rocm_tracer.cc

```diff
@@ -0,0 +1,1132 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/profiler/internal/gpu/rocm_tracer.h"
+
+#include <chrono>
+#include <iostream>
+#include <sstream>
+#include <thread>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/node_hash_map.h"
+#include "rocm/rocm_config.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/lib/gtl/cleanup.h"
+#include "tensorflow/core/lib/hash/hash.h"
+#include "tensorflow/core/platform/env.h"
+#include "tensorflow/core/platform/logging.h"
+#include "tensorflow/core/platform/macros.h"
+#include "tensorflow/core/platform/mem.h"
+#include "tensorflow/core/profiler/internal/cpu/annotation_stack.h"
+
+namespace tensorflow {
+namespace profiler {
+
+constexpr uint32_t RocmTracerEvent::kInvalidDeviceId;
+
+#define RETURN_IF_ROCTRACER_ERROR(expr)                                      \
+  do {                                                                       \
+    roctracer_status_t status = expr;                                        \
+    if (status != ROCTRACER_STATUS_SUCCESS) {                                \
+      const char* errstr = wrap::roctracer_error_string();                   \
+      LOG(ERROR) << "function " << #expr << "failed with error " << errstr;  \
+      return errors::Internal(absl::StrCat("roctracer call error", errstr)); \
+    }                                                                        \
+  } while (false)
+
+namespace {
+
+// GetCachedTID() caches the thread ID in thread-local storage (which is a
+// userspace construct) to avoid unnecessary system calls. Without this caching,
+// it can take roughly 98ns, while it takes roughly 1ns with this caching.
+int32_t GetCachedTID() {
+  static thread_local int32_t current_thread_id =
+      Env::Default()->GetCurrentThreadId();
+  return current_thread_id;
+}
+
+const char* GetActivityDomainName(uint32_t domain) {
+  switch (domain) {
+    case ACTIVITY_DOMAIN_HSA_API:
+      return "HSA API";
+    case ACTIVITY_DOMAIN_HSA_OPS:
+      return "HSA OPS";
+    // case ACTIVITY_DOMAIN_HIP_OPS:
+    //   return "HIP OPS";
+    case ACTIVITY_DOMAIN_HCC_OPS:
+      return "HCC OPS";
+    // case ACTIVITY_DOMAIN_HIP_VDI:
+    //   return "HIP VDI";
+    case ACTIVITY_DOMAIN_HIP_API:
+      return "HIP API";
+    case ACTIVITY_DOMAIN_KFD_API:
+      return "KFD API";
+    case ACTIVITY_DOMAIN_EXT_API:
+      return "EXT API";
+    case ACTIVITY_DOMAIN_ROCTX:
+      return "ROCTX";
+    default:
+      DCHECK(false);
+      return "";
+  }
+  return "";
+}
+
+string GetActivityDomainOpName(uint32_t domain, uint32_t op) {
```

Comment:
```suggestion
std::string GetActivityDomainOpName(uint32_t domain, uint32_t op) {
```

One recent development is that we're qualifying std std::string everywhere. :)

Reply:
changed to `std::string` in the latest commit